fish-rust/src/curses.rs

```diff
@@ -36,11 +36,32 @@ pub fn term() -> Option<Arc<Term>> {
         .map(Arc::clone)
 }
 
+/// Convert a nul-terminated pointer, which must not be itself null, to a CString.
+fn ptr_to_cstr(ptr: *const libc::c_char) -> CString {
+    assert!(!ptr.is_null());
+    unsafe { CStr::from_ptr(ptr).to_owned() }
+}
+
+/// Convert a nul-terminated pointer to a CString, or None if the pointer is null.
+fn try_ptr_to_cstr(ptr: *const libc::c_char) -> Option<CString> {
+    if ptr.is_null() {
+        None
+    } else {
+        Some(ptr_to_cstr(ptr))
+    }
+}
+
 /// Private module exposing system curses ffi.
 mod sys {
     pub const OK: i32 = 0;
     pub const ERR: i32 = -1;
 
+    /// tputs callback argument type and the callback type itself.
+    /// N.B. The C++ had a check for TPUTS_USES_INT_ARG for the first parameter of tputs
+    /// which was to support OpenIndiana, which used a char.
```

Comment:
It's not so much that the libc crate shields us from anything but that we are using the binary C abi directly which is type-agnostic/oblivious. I wouldn't trust the `libc` crate definition to actually match the underlying for esoteric platforms; the question is what difference is there in the (primarily x86_64) calling convention when passing a char argument vs when passing an int argument. If it was signed vs unsigned we could ignore that altogether but since one is 1 byte and the other is 4 bytes, they actually end up being treated differently in the function preamble at the call site and the function prologue in the callee.

The const stuff is truly vanquished as that's just a type system artifact and doesn't survive in any way, shape, or form to the underlying assembly (optimizations excepted).

Reply:
Yeah that's right. This falls into the bucket of potentially breaking niche platforms for the duration of the port. We'll support OpenIndiana if feasible, but after the port is done.